commerce-cloud-code/app_yourshophere/cartridge/controllers/Cart.js

```
@@ -67,7 +83,171 @@ server.use('Delete', (req, res, next) => {
+    res.redirect(redirectUrl.append('success', Resource.msg('info.shipping.updated', 'cart', null)));
+    return next();
+});
+
+server.post('AddCoupon', server.middleware.https, (req, res, next) => {
+    const BasketMgr = require('dw/order/BasketMgr');
+    const URLUtils = require('dw/web/URLUtils');
+    const Transaction = require('dw/system/Transaction');
+
+    const CouponStatusCodes = require('dw/campaign/CouponStatusCodes');
+    const currentBasket = BasketMgr.getCurrentOrNewBasket();
+    const couponCode = req.httpParameterMap.couponCode.stringValue;
+    const redirectUrl = request.httpParameterMap.hx.submitted ? URLUtils.url('Cart-Show', 'hx', request.httpParameterMap.hx.stringValue || '') : URLUtils.url('Cart-Show');
+
+    if (!currentBasket) {
+        // Could log this error for monitoring
+    }
+
+    if (!couponCode) {
+        // Could log this error
+    }
+
+    try {
+        Transaction.wrap(() => {
+            currentBasket.createCouponLineItem(couponCode, true);
+        });
+    } catch (e) {
+        if (e.errorCode === CouponStatusCodes.COUPON_CODE_ALREADY_IN_BASKET) {
+            // ...
+        }
+    }
+
+    // If successful, redirect to the cart page
+    res.redirect(redirectUrl);
+
+    next();
+});
+
+server.get('RemoveCoupon', server.middleware.https, (req, res, next) => {
+    const BasketMgr = require('dw/order/BasketMgr');
+    const URLUtils = require('dw/web/URLUtils');
+    const Transaction = require('dw/system/Transaction');
+
+    const CouponStatusCodes = require('dw/campaign/CouponStatusCodes');
+    const currentBasket = BasketMgr.getCurrentOrNewBasket();
+    const index = req.httpParameterMap.index.intValue || 0;
+    const redirectUrl = request.httpParameterMap.hx.submitted ? URLUtils.url('Cart-Show', 'hx', request.httpParameterMap.hx.stringValue || '') : URLUtils.url('Cart-Show');
+
+    if (currentBasket.couponLineItems.length <= index) {
+        // Could log this error
```

please throw an error for now so it is obvious error handling need to be build. This silent treatment would potentially cause long lasting headscratching and debugging